hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java

```diff
@@ -0,0 +1,288 @@
+   * to initialize it's state, and perform any costly calculation.
+   */
+  @Override
+  void init(final BaseLoadBalancer.Cluster cluster) {
+    this.cluster = cluster;
+    this.loadRules();
+  }
+
+  @Override
+  protected double cost() {
+
+    double cost = 0;
+
+    final double targetUsage = ((double) this.cluster.numRegions / (double) this.totalCapacity);
+
+    for (int i = 0; i < this.cluster.numServers; i++) {
+
+      // retrieve capacity for each RS
+      final ServerName sn = this.cluster.servers[i];
+      final double limit = this.findLimitForRS(sn);
+      final double nbrRegions = this.cluster.regionsPerServer[i].length;
+
+      // cost is the number of regions above the local limit
+      final double usage = nbrRegions / limit;
+      if (usage > targetUsage) {
+        // cost is the number of regions above the local limit
+        final double localCost = (nbrRegions - Math.round(limit * targetUsage)) / limit;
+        cost += localCost;
+      }
+    }
+    return cost / (double) this.cluster.numServers;
+
+  }
+
+  /**
+   * used to load the rule files.
+   */
+  void loadRules() {
+    final List<String> lines = HeterogeneousRegionCountCostFunction.readFile(this.rulesPath);
+    if (lines.size() == 0) {
+      return;
+    }
+
+    LOG.info("loading rules file '" + this.rulesPath + "'");
+    this.limitPerRule.clear();
+    for (final String line : lines) {
+      try {
+        if (line.length() == 0) {
+          continue;
+        }
+
+        if (line.startsWith("#")) {
+          continue;
+        }
+
+        final String[] splits = line.split(" ");
+        if (splits.length != 2) {
+          throw new IOException("line '" + line + "' is malformated, " +
+                  "expected [regexp] [limit]. Skipping line");
+        }
+
+        final Pattern pattern = Pattern.compile(splits[0]);
+        final Integer limit = Integer.parseInt(splits[1]);
+        this.limitPerRule.put(pattern, limit);
+      } catch (final IOException | NumberFormatException | PatternSyntaxException e) {
+        LOG.error("error on line: " + e);
+      }
+    }
+
+    this.rebuildCache();
+  }
+
+  /**
+   * used to read the rule files from either HDFS or local FS
+   */
+  private static List<String> readFile(final String filename) {
+    final List<String> records = new ArrayList<String>();
+
+    if (null == filename) {
+      return records;
+    }
+
+    final Configuration conf = new Configuration();
+    final Path path = new Path(filename);
+    try {
+      final FileSystem fs = FileSystem.get(conf);
+      final BufferedReader reader = new BufferedReader(new InputStreamReader(fs.open(path)));
+
+      String line;
+      while ((line = reader.readLine()) != null) {
+        records.add(line);
+      }
+      reader.close();
+      return records;
+    } catch (final IOException e) {
+      LOG.error("cannot read rules file located at ' " + filename + " ':" + e.getMessage());
+    }
+    return records;
+  }
```

Comment:
This reads from HDFS only. Will read from local FS if HDFS is configured in local mode, instead of distributed, but that is only reasonable for tests purposes.

Reply:
In my opinion, this should work on both cases, depending on the prefix. The current implementation is working on local, not tested with HDFS

Reply:
It will work when using hbase with local file system mode, but not when running in distributed mode. I just confirmed that by trying similar code on a cluster in distributed mode:

`WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
java.lang.IllegalArgumentException: Wrong FS: file:/root/test1.txt, expected: hdfs://mycluster.com:8020
	at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:662)`

Anyways, I think it's fine to leave it reading from HDFS only, we would just need to adjust the comments in this method and in the class description.